### PR TITLE
Add various time and date functions

### LINKED ISSUE #1545: Day/month functions
I have few workflows when I need to check metrics over specific days.

It would be nice to have few functions which return a day of the week, day of the year, month etc.

examples:
```
day_of_the_week(time()) == 2
month(time()) == 12
```

### PR DESCRIPTION
**Pull request by brian-brazil**
Adds day_of_month, day_of_week, hour_of_day, month_of_year, year and days_in_month.

This all take in unix timestamps as a vector, and return the answer as a vector.
All results are in UTC.

Fixes #1545

### CODE REVIEW

File: promql/functions.go [hunk 1]
```diff
@@ -1010,6 +1097,12 @@ var functions = map[string]*Function{
 		ReturnType: model.ValVector,
 		Call:       funcMinOverTime,
 	},
+	"month_of_year": {
```
Comment (grobie) [inline promql/functions.go]: Why is it month of year and not just month?

Comment (brian-brazil) [inline promql/functions.go]: Someone could ask for month of century etc. Better to be explicit.

Comment (grobie) [inline promql/functions.go]: This is a very far fetched example without any real use in practice. I'm against adding support for such exotic date functions to the query language. The definition of hour/day/month/year is very clear and established, there is no need to be overly explicit. Usability has to be kept in mind as well and long function names are not helpful when reading expressions.

I vote to rename hour_of_day() to hour() and month_of_year() to month(). The day_* functions make sense, even though I'd prefer day() over day_of_month().

Comment (fabxc) [inline promql/functions.go]: Agreed. These are the absolutely sane default cases. If we were to add others, we can then extend by suffixes.

Comment (brian-brazil) [inline promql/functions.go]: PR updated.

**Comment by grobie** [comment]
So `time()` returns a scalar, but all the date functions expect a vector. I assume using these date functions for the current time is the most common use case for these functions, which requires a rather laborious `month(vector(time()))`.

I understand that it's desired to be able to use the date functions on vector results. So my only idea would be to make the argument optional and use `vector(time())` as default value. What do you think? Other suggestions?

**Comment by tcolgate** [comment]
For my "predict_linear" to end of month, I'd need something like
```
predict_linear(...,scalar(days_in_month() - day_of_month()) * 86400 + (86400 - (time() % 86400))
```
It's not exactly pretty, but not unreasonable.

**Comment by fabxc** [comment]
Well, we could make the functions accept either type or, as you suggested, have a default. I've voted strongly against such dynamic typing in the past so I'm torn whether I should give up because it seems convenient in this one case.

In general, both solutions would be additions and could be added later after the functions have some milage as they are now.

**Comment by brian-brazil** [comment]
I'm also strongly against dynamic typing. I'm generally against optional arguments to keep functions understandable, but it may make sense here.

Waiting and seeing is a safe path.

**Comment by beorn7** [comment]
The default value approach would still imply static typing (as the default value is a vector, too).

**Comment by grobie** [comment]
For what would we be waiting? All presented use cases were for current
time() so far.

**Comment by tcolgate** [comment]
My example without the defaulting is
`predict_linear(...,scalar(days_in_month(vector(time())) - day_of_month(vector(time()))) * 86400 + (86400 - (time() % 86400))`

I'm struggling a bit to think of the use case for that not being the current time. Maybe if I was getting the time from another metric (like last successful completion of a job), but then I'd mostly care about an absolute number of days. I don't see a strong need to determine if my last job failure was on a specific week day (I might want to check if it has just failed, and it's a saturday, maybe)

The optional argument makes a lot of sense.

**Comment by brian-brazil** [comment]
I'd added the default.

**Comment by fabxc** [comment]
Mh, not sure I understand. Didn't we both say above we are against dynamic typing?

**Comment by brian-brazil** [comment]
There's no dynamic typing here, the functions take a vector and return a vector.

**Comment by grobie** [comment]
:+1: While I'd still rename `day_of_month()` to `day()` given this is the most common usage and matches what most other programming languages do. `weekday()` used by golang is also shorter than `day_of_week()`.

**Comment by fabxc** [comment]
Yes, and you just made it so that you can omit an argument. This is no static typing anymore and we haven't specified whether we want to support variadic arguments.

**Comment by brian-brazil** [comment]
That's still static typing, and we do support optional arguments already.

**Comment by fabxc** [comment]
And we had discussions that it's not a good idea and we should get rid of them.

**Comment by beorn7** [comment]
IIRC the case where we got rid of default values for arguments was `delta`, and that was not so much because it had a default argument, but because it was a difficult to read argument (what does "1" mean here?) and a boolean flag (which usually translates better into a different function name, which we introduced).

In this case, the default value for the argument is quite the opposite: It improves readability, and it is not a boolean flag. Alternatives would be to introduce a separate set of functions (double the number, have unwieldy names for them) or to make the sure write `vector(time())` in almost all the use cases.

I wouldn't subscribe to the notion that default arguments are dynamic typing. But that's more an academic discussion. The reason why C or Go don't have them, are manifold, and certainly different for each of those languages. As you can define your own functions in those languages, the comparison to PromQL, where you cannot, has to be taken with a grain of salt.

**Comment by grobie** [comment]
All other optional arguments I saw had a very different context. But given the only feature requests are for date fuctions of the current time, could we just postpone this discussion and not offer any argument for now?

**Comment by grobie** [comment]
I didn't see @beorn7's comment, but agree on all points.

**Comment by fabxc** [comment]
As I said, I don't necessarily disagree this being applied here. But seems like we prefer arguing terminology here rather than long-term QL effects.

Whatever, seems like general interest is getting this in anyway, so 👍

**Pull request merged**